.github/CONTRIBUTING.md

```diff
@@ -0,0 +1,89 @@
+
+The following are examples of good commit messages:
+
+- `chore: renamed nnid service to nnas`
+- `bug-fix: fixed hang in MutexMap.Has`
+
+Please refer to [Conventional Commits](https://conventionalcommits.org/) for a detailed guide on how to structure commit messages. Writing good, detailed, commit messages helps ensure that we can refer back to the git history and quickly find where specific changes occurred in the event that they need further review, reverting, etc.
+
+# Pull Requests
+As stated in [Making Issues](#making-issues), before making a pull request there must be an open, approved, issue for the changes being made. If you can not find an approved issue for the changes you wish to make, please make one before continuing.
+
+Before making a pull request ensure you have tested all changes and that no regressions have been introduced.
+
+Pull requests should never be made against the default (`main`/`master`) branch of a repository. The default branch contains the most recent, stable, version of the codebase. All work on the codebase should take place in other branches. Unless otherwise specified, your target branch should typically be the `dev` branch. You may target other feature branches, however, if need be. If a `dev` branch does not exist for the repository you are working on, please submit a feature request for one to be added before continuing.
+
+Like everything else, pull requests should be as detailed as possible. Your title should adequately summarize the changes being made, and the body of the pull request should fully explain your changes. We ask that, if applicable, the rationale behind your changes also be noted. For example rather than simply "Changed from `for...of` to `forEach`", if the change was made for a performance reason you should say "Changed from `for...of` to `forEach` due to `forEach` being X times faster in this case" and provide some benchmarks. Adding images, videos, etc. is also welcomed in order to illustrate changes. If the changes being made are directly tied to some form of visual (such as a change to the website, a tools GUI, etc.) then images or videos is ***REQUIRED***. If none are provided, then we may delay review until they are given. Providing visual examples of these changes allow us to quickly assess whether or not we wish to proceed with the changes being made.
```

Comment:
It's not obvious how to benchmark most of our stuff - e.g. NEX servers under high load - so is this actually a case for us to write up test harnesses? Is that a good use of our time right now?

Reply:
Right, so providing benchmarks for performance changes isn't a rule, that was an example of providing details for evaluation.

Reply:
It was an example of how to explain the reasoning behind the changes. Nothing more and nothing less. I'm not sure what the issue is with this?

The point was to explain why changes were made if applicable. The example being changes made for performance reasons. If someone makes a change for performance reasons, it stands to reason that in that example it's because of something they could be measured (otherwise how would the performance issue even be found and known to be fixed). Thus in the example you could provide a benchmark for the changes

Again, I'm not understanding the issue here? It was literally just an example situation that I made up to illustrate a point

Reply:
Yup, no issue. Was just making sure I understood what you said correctly.

Reply:
My apologies then! I thought you still had some sort of concern about it, I misunderstood